core/service/src/lib.rs

```diff
@@ -525,6 +551,50 @@ impl<Components> Service<Components> where Components: components::Components {
 	}
 }
 
+impl<Components> Future for Service<Components> where Components: components::Components {
+	type Item = ();
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		Future::poll(&mut &*self)
+	}
+}
+
+// Note that this implementation is totally unnecessary. It exists only because of tests. The tests
+// should eventually be reworked, as it would make it possible to remove the `Mutex`es. that we
+// lock here.
+impl<'a, Components> Future for &'a Service<Components> where Components: components::Components {
+	type Item = ();
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		// The user is supposed to poll only one service, so it doesn't matter if we keep this
+		// mutex locked.
+		let mut to_poll = self.to_poll.lock();
+		let mut to_spawn_rx = self.to_spawn_rx.lock();
+
+		while let Ok(Async::Ready(Some(task_to_spawn))) = to_spawn_rx.poll() {
+			let executor = tokio_executor::DefaultExecutor::current();
+			if let Err(err) = executor.execute(task_to_spawn) {
```

Comment:
Can we not do this check once?

Reply:
No, theoretically the executor could have a limited capacity, and so later tasks can fail to spawn.